app/controllers/mixins/ems_common_angular.rb

```diff
@@ -77,11 +75,53 @@ def create
 
       case params[:button]
       when "add" then create_ems_button_add
-      when "validate" then create_ems_button_validate
+      when "validate" then update_ems_button_validate
       when "cancel" then create_ems_button_cancel
       end
     end
 
+    def queue_authentication_check(ems_type)
+      task_opts = {
+        :action => "Validate EMS Provider Credentials",
+        :userid => session[:userid]
+      }
+
+      queue_opts = {
+        :args => [*get_task_args(ems_type)],
+        :class_name => ems_type,
+        :method_name => "raw_connect",
+        :queue_name => "generic",
+        :role => "ems_operations",
+        :zone => params[:zone]
+      }
+
+      task_id = MiqTask.generic_action_with_callback(task_opts, queue_opts)
+      task = MiqTask.wait_for_taskid(task_id, :timeout => 30)
+
+      if task.nil?
+        error_message = "Task Error"
+      elsif MiqTask.status_error?(task.status) || MiqTask.status_timeout?(task.status)
+        error_message = task.message
+      end
+
+      [!error_message.present?, error_message]
+    end
+
+    def get_task_args(ems)
+      if ems == ManageIQ::Providers::Openstack::CloudManager
+        auth_url = ems.class.auth_url(params[:default_hostname], params[:default_api_port])
+        [params[:default_userid], params[:default_password], auth_url]
+      elsif ems == ManageIQ::Providers::Amazon::CloudManager
+        [params[:default_userid], params[:default_password], :EC2, params[:provider_region], true]
+      elsif ems == ManageIQ::Providers::Azure::CloudManager
+        [params[:default_userid], params[:default_password], params[:azure_tenant_id], params[:subscription], params[:proxy_uri], params[:provider_region]]
+      elsif ems == ManageIQ::Providers::Vmware::CloudManager
+        [params[:default_hostname], params[:default_api_port], params[:default_userid], params[:default_password], true]
+      elsif ems == ManageIQ::Providers::Google::CloudManager
+        [params[:project], params[:service_account], {:service => "compute"}]
+      end
+    end
+
```

Comment:
Looks like a good opportunity for case/when. Also, should there be an "else" condition?